main/src/ca/uwaterloo/flix/api/lsp/provider/CodeActionProvider.scala

```diff
@@ -187,18 +188,54 @@ object CodeActionProvider {
     * Returns a code action that proposes to `use` a type.
     */
   private def mkUseType(ident: Name.Ident, uri: String, ap: AnchorPosition)(implicit root: Root): List[CodeAction] = {
-    val names = root.enums.map { case (sym, _) => sym.name } ++
-      root.restrictableEnums.map { case (sym, _) => sym.name } ++
-      root.traits.map { case (sym, _) => sym.name } ++
-      root.typeAliases.map { case (sym, _) => sym.name }
+    val enumNames = root.enums.collect { case (sym, enm) if enumIsPublic(enm) => sym.name }
+    val enumSyms = root.enums.collect { case (sym, enm) if enumIsPublic(enm) => sym }
 
-    val syms = (root.enums ++ root.restrictableEnums ++ root.traits ++ root.typeAliases).map {
-      case (sym, _) => sym
-    }
+    val restrictableEnumNames = root.restrictableEnums.collect { case (sym, enm) if restrictableEnumIsPublic(enm) => sym.name }
+    val restrictableEnumSyms = root.restrictableEnums.collect { case (sym, enm) if restrictableEnumIsPublic(enm) => sym }
+
+    val traitNames = root.traits.collect { case (sym, trt) if traitIsPublic(trt) => sym.name }
+    val traitSyms = root.traits.collect { case (sym, trt) if traitIsPublic(trt) => sym }
+
+    val typeAliasNames = root.typeAliases.collect { case (sym, alias) if typeAliasIsPublic(alias) => sym.name }
+    val typeAliasSyms = root.typeAliases.collect { case (sym, alias) if typeAliasIsPublic(alias) => sym }
+
+    val names = enumNames ++ restrictableEnumNames ++ traitNames ++ typeAliasNames
+    val syms = enumSyms ++ restrictableEnumSyms ++ traitSyms ++ typeAliasSyms
 
     mkUseSym(ident, names, syms, uri, ap)
   }
 
+  /**
+    * Checks if the given spec is public.
+    */
+  private def specIsPublic(spec: TypedAst.Spec): Boolean = spec.mod.mod.contains(Modifier.Public)
+
+  /**
+    * Checks if the given trait is public.
+    */
+  private def traitIsPublic(trt: TypedAst.Trait): Boolean = trt.mod.mod.contains(Modifier.Public)
+
+  /**
+    * Checks if the given effect is public.
+    */
+  private def effectIsPublic(eff: TypedAst.Effect): Boolean = eff.mod.mod.contains(Modifier.Public)
+
+  /**
+    * Checks if the given enum is public.
+    */
+  private def enumIsPublic(enm: TypedAst.Enum): Boolean = enm.mod.mod.contains(Modifier.Public)
+
+  /**
+    * Checks if the given restrictable enum is public.
+    */
+  private def restrictableEnumIsPublic(enm: TypedAst.RestrictableEnum): Boolean = enm.mod.mod.contains(Modifier.Public)
+
+  /**
+    * Checks if the given type alias is public.
+    */
+  private def typeAliasIsPublic(alias: TypedAst.TypeAlias): Boolean = alias.mod.mod.contains(Modifier.Public)
```

Comment:
Lets rename all of the above to just `isPublic` and rely on overloading to pick the right one.